diff --git a/Telegram/CMakeLists.txt b/Telegram/CMakeLists.txt
index 30ddb5a167..41db7d7a63 100644
--- a/Telegram/CMakeLists.txt
+++ b/Telegram/CMakeLists.txt
@@ -103,6 +103,9 @@ nice_target_sources(Telegram ${src_loc}
 PRIVATE
     ${style_files}
 
+    blorg/discovery/config.h
+    blorg/discovery/discovery.cpp
+    blorg/discovery/discovery.h
     api/api_attached_stickers.cpp
     api/api_attached_stickers.h
     api/api_authorizations.cpp
@@ -2257,7 +2260,7 @@ else()
 endif()
 
 if (build_macstore)
-    set(bundle_identifier "org.telegram.desktop")
+    set(bundle_identifier "blorg.telegram.desktop")
     set(bundle_entitlements "Telegram Lite.entitlements")
     set(output_name "Telegram Lite")
     target_link_options(Telegram
@@ -2280,7 +2283,7 @@ else()
         set(bundle_identifier "com.tdesktop.Telegram")
     endif()
     set(bundle_entitlements "Telegram.entitlements")
-    set(output_name "Telegram")
+    set(output_name "TelegramBlorg")
 endif()
 
 if (CMAKE_GENERATOR STREQUAL Xcode)
@@ -2502,28 +2505,28 @@ endif()
 
 if (LINUX AND DESKTOP_APP_USE_PACKAGED)
     include(GNUInstallDirs)
-    configure_file("../lib/xdg/org.telegram.desktop.service" "${CMAKE_CURRENT_BINARY_DIR}/org.telegram.desktop.service" @ONLY)
-    configure_file("../lib/xdg/org.telegram.desktop.metainfo.xml" "${CMAKE_CURRENT_BINARY_DIR}/org.telegram.desktop.metainfo.xml" @ONLY)
-    generate_appstream_changelog(Telegram "${CMAKE_SOURCE_DIR}/changelog.txt" "${CMAKE_CURRENT_BINARY_DIR}/org.telegram.desktop.metainfo.xml")
+    configure_file("../lib/xdg/blorg.telegram.desktop.service" "${CMAKE_CURRENT_BINARY_DIR}/blorg.telegram.desktop.service" @ONLY)
+    configure_file("../lib/xdg/blorg.telegram.desktop.metainfo.xml" "${CMAKE_CURRENT_BINARY_DIR}/blorg.telegram.desktop.metainfo.xml" @ONLY)
+    generate_appstream_changelog(Telegram "${CMAKE_SOURCE_DIR}/changelog.txt" "${CMAKE_CURRENT_BINARY_DIR}/blorg.telegram.desktop.metainfo.xml")
     install(TARGETS Telegram RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}" BUNDLE DESTINATION "${CMAKE_INSTALL_BINDIR}")
-    install(FILES "Resources/art/icon16.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/16x16/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon16@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/16x16@2/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon32.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/32x32/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon32@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/32x32@2/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon48.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/48x48/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon48@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/48x48@2/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon64.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/64x64/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon64@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/64x64@2/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon128.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/128x128/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon128@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/128x128@2/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon256.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/256x256/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon256@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/256x256@2/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon512.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/512x512/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/art/icon512@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/512x512@2/apps" RENAME "org.telegram.desktop.png")
-    install(FILES "Resources/icons/tray_monochrome.svg" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/symbolic/apps" RENAME "org.telegram.desktop-symbolic.svg")
-    install(FILES "Resources/icons/tray_monochrome_attention.svg" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/symbolic/apps" RENAME "org.telegram.desktop-attention-symbolic.svg")
-    install(FILES "Resources/icons/tray_monochrome_mute.svg" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/symbolic/apps" RENAME "org.telegram.desktop-mute-symbolic.svg")
-    install(FILES "../lib/xdg/org.telegram.desktop.desktop" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/applications")
-    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/org.telegram.desktop.service" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/dbus-1/services")
-    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/org.telegram.desktop.metainfo.xml" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/metainfo")
+    install(FILES "Resources/art/icon16.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/16x16/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon16@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/16x16@2/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon32.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/32x32/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon32@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/32x32@2/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon48.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/48x48/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon48@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/48x48@2/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon64.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/64x64/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon64@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/64x64@2/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon128.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/128x128/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon128@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/128x128@2/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon256.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/256x256/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon256@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/256x256@2/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon512.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/512x512/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/art/icon512@2x.png" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/512x512@2/apps" RENAME "blorg.telegram.desktop.png")
+    install(FILES "Resources/icons/tray_monochrome.svg" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/symbolic/apps" RENAME "blorg.telegram.desktop-symbolic.svg")
+    install(FILES "Resources/icons/tray_monochrome_attention.svg" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/symbolic/apps" RENAME "blorg.telegram.desktop-attention-symbolic.svg")
+    install(FILES "Resources/icons/tray_monochrome_mute.svg" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/symbolic/apps" RENAME "blorg.telegram.desktop-mute-symbolic.svg")
+    install(FILES "../lib/xdg/blorg.telegram.desktop.desktop" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/applications")
+    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/blorg.telegram.desktop.service" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/dbus-1/services")
+    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/blorg.telegram.desktop.metainfo.xml" DESTINATION "${CMAKE_INSTALL_DATAROOTDIR}/metainfo")
 endif()
diff --git a/Telegram/Resources/qrc/telegram/telegram.qrc b/Telegram/Resources/qrc/telegram/telegram.qrc
index 2aac736dbc..618ecb1c99 100644
--- a/Telegram/Resources/qrc/telegram/telegram.qrc
+++ b/Telegram/Resources/qrc/telegram/telegram.qrc
@@ -67,6 +67,6 @@
   </qresource>
   <qresource prefix="/misc">
     <file alias="default_shortcuts-custom.json">../../default_shortcuts-custom.json</file>
-    <file alias="org.telegram.desktop.desktop">../../../../lib/xdg/org.telegram.desktop.desktop</file>
+    <file alias="blorg.telegram.desktop.desktop">../../../../lib/xdg/blorg.telegram.desktop.desktop</file>
   </qresource>
 </RCC>
diff --git a/Telegram/SourceFiles/api/api_chat_invite.cpp b/Telegram/SourceFiles/api/api_chat_invite.cpp
index 95bd8d7687..8c96b654dd 100644
--- a/Telegram/SourceFiles/api/api_chat_invite.cpp
+++ b/Telegram/SourceFiles/api/api_chat_invite.cpp
@@ -44,6 +44,9 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "styles/style_credits.h"
 #include "styles/style_info.h"
 #include "styles/style_layers.h"
+#include "styles/style_premium.h"
+#include "blorg/discovery/discovery.h"
+
 
 namespace Api {
 
@@ -442,6 +445,26 @@ void ConfirmInviteBox(
 	const auto isChannel = invite.isChannel && !invite.isMegagroup;
 	const auto requestApprove = invite.isRequestNeeded;
 	const auto count = invite.participantsCount;
+	
+	if (!BLORG::Discovery::AllowRequestsJoin() && requestApprove) {
+        box->setNoContentMargin(true);
+        box->setWidth(st::boxWideWidth);
+        const auto content = box->verticalLayout();
+
+        Ui::AddSkip(content, st::confirmInvitePhotoTop);
+        
+        box->addRow(
+            object_ptr<Ui::FlatLabel>(
+                box,
+                QString("Entry to channels with requests is blocked"),
+                st::confirmInviteTitle),
+            style::al_top);
+
+        Ui::AddSkip(content, st::boxPadding.bottom());
+
+        box->addButton(tr::lng_close(), [=] { box->closeBox(); });
+        return;
+    }
 
 	struct State {
 		std::shared_ptr<Data::PhotoMedia> photoMedia;
diff --git a/Telegram/SourceFiles/api/api_peer_search.cpp b/Telegram/SourceFiles/api/api_peer_search.cpp
index 73a252ab12..b5bda994be 100644
--- a/Telegram/SourceFiles/api/api_peer_search.cpp
+++ b/Telegram/SourceFiles/api/api_peer_search.cpp
@@ -12,6 +12,7 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "data/data_session.h"
 #include "dialogs/ui/chat_search_in.h" // IsHashOrCashtagSearchQuery
 #include "main/main_session.h"
+#include "blorg/discovery/discovery.h"
 
 namespace Api {
 namespace {
@@ -67,6 +68,10 @@ void PeerSearch::requestPeers() {
 		MTP_string(_query),
 		MTP_int(SearchPeopleLimit)
 	)).done([=](const MTPcontacts_Found &result, mtpRequestId requestId) {
+		if (!BLORG::Discovery::AllowGlobalSearch()) {
+			return;
+		}
+
 		const auto &data = result.data();
 		_session->data().processUsers(data.vusers());
 		_session->data().processChats(data.vchats());
diff --git a/Telegram/SourceFiles/blorg/discovery/config.h b/Telegram/SourceFiles/blorg/discovery/config.h
new file mode 100644
index 0000000000..3f0b52ff42
--- /dev/null
+++ b/Telegram/SourceFiles/blorg/discovery/config.h
@@ -0,0 +1,13 @@
+#pragma once
+
+namespace BLORG::Discovery {
+
+inline constexpr bool kBlockGlobalSearch = true;
+inline constexpr bool kBlockUsernameResolve = true;
+inline constexpr bool kBlockTelegramLinks = true;
+inline constexpr bool kBlockInlineBots = true;
+inline constexpr bool kBlockBlockedUsersSection = true;
+inline constexpr bool kBlockRequestsJoin = true;
+inline constexpr bool kBlockForwardedMessages = true;
+
+}
\ No newline at end of file
diff --git a/Telegram/SourceFiles/blorg/discovery/discovery.cpp b/Telegram/SourceFiles/blorg/discovery/discovery.cpp
new file mode 100644
index 0000000000..043bba0fb3
--- /dev/null
+++ b/Telegram/SourceFiles/blorg/discovery/discovery.cpp
@@ -0,0 +1,40 @@
+#include <QString>
+#include "blorg/discovery/discovery.h"
+#include "blorg/discovery/config.h"
+
+namespace BLORG::Discovery {
+
+bool AllowMentions() {
+    return !kBlockUsernameResolve;
+}
+
+bool AllowInlineBots() {
+    return !kBlockInlineBots;
+}
+
+bool AllowTelegramLinks() {
+    return !kBlockTelegramLinks;
+}
+
+bool AllowResolveUsername() {
+    return !kBlockUsernameResolve;
+}
+
+bool AllowGlobalSearch() {
+    return !kBlockGlobalSearch;
+}
+
+bool ShowBlockedUsersSection() {
+    return !kBlockBlockedUsersSection;
+}
+
+bool AllowRequestsJoin() {
+    return !kBlockRequestsJoin;
+}
+
+bool AllowForwardedMessages() {
+    return !kBlockForwardedMessages;
+
+}
+
+}
\ No newline at end of file
diff --git a/Telegram/SourceFiles/blorg/discovery/discovery.h b/Telegram/SourceFiles/blorg/discovery/discovery.h
new file mode 100644
index 0000000000..356f39f25d
--- /dev/null
+++ b/Telegram/SourceFiles/blorg/discovery/discovery.h
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <QString>
+
+namespace BLORG::Discovery {
+
+bool AllowMentions();
+
+bool AllowTelegramLinks();
+
+bool AllowResolveUsername();
+
+bool AllowInlineBots();
+
+bool AllowGlobalSearch();
+
+bool ShowBlockedUsersSection();
+
+bool AllowRequestsJoin();
+
+bool AllowForwardedMessages();
+
+}
\ No newline at end of file
diff --git a/Telegram/SourceFiles/chat_helpers/message_field.cpp b/Telegram/SourceFiles/chat_helpers/message_field.cpp
index 0382ee6768..6ed9c73204 100644
--- a/Telegram/SourceFiles/chat_helpers/message_field.cpp
+++ b/Telegram/SourceFiles/chat_helpers/message_field.cpp
@@ -58,6 +58,7 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "styles/style_premium.h"
 #include "styles/style_settings.h"
 #include "base/qt/qt_common_adapters.h"
+#include "blorg/discovery/discovery.h"
 
 #include <QtCore/QMimeData>
 #include <QtCore/QStack>
@@ -835,6 +836,10 @@ InlineBotQuery ParseInlineBotQuery(
 	const auto &full = field->getTextWithTags();
 	const auto &text = full.text;
 	const auto textLength = text.size();
+	
+	if (!BLORG::Discovery::AllowInlineBots()) {
+		return result;
+	}
 
 	auto inlineUsernameStart = 1;
 	auto inlineUsernameLength = 0;
diff --git a/Telegram/SourceFiles/core/application.cpp b/Telegram/SourceFiles/core/application.cpp
index 22fcbbdb35..ac792c8385 100644
--- a/Telegram/SourceFiles/core/application.cpp
+++ b/Telegram/SourceFiles/core/application.cpp
@@ -1938,7 +1938,7 @@ void Application::RegisterUrlScheme() {
 		.arguments = arguments,
 		.protocol = u"tg"_q,
 		.protocolName = u"Telegram Link"_q,
-		.shortAppName = u"tdesktop"_q,
+		.shortAppName = u"telegramblorg"_q,
 		.longAppName = QCoreApplication::applicationName(),
 		.displayAppName = AppName.utf16(),
 		.displayAppDescription = AppName.utf16(),
@@ -1949,7 +1949,7 @@ void Application::RegisterUrlScheme() {
 		.arguments = arguments,
 		.protocol = u"tonsite"_q,
 		.protocolName = u"TonSite Link"_q,
-		.shortAppName = u"tdesktop"_q,
+		.shortAppName = u"telegramblorg"_q,
 		.longAppName = QCoreApplication::applicationName(),
 		.displayAppName = AppName.utf16(),
 		.displayAppDescription = AppName.utf16(),
diff --git a/Telegram/SourceFiles/core/launcher.cpp b/Telegram/SourceFiles/core/launcher.cpp
index 2487d4d5bd..e8e98cf229 100644
--- a/Telegram/SourceFiles/core/launcher.cpp
+++ b/Telegram/SourceFiles/core/launcher.cpp
@@ -336,7 +336,7 @@ void Launcher::init() {
 	prepareSettings();
 	initQtMessageLogging();
 
-	QApplication::setApplicationName(u"TelegramDesktop"_q);
+	QApplication::setApplicationName(u"TelegramBlorg"_q);
 
 #if QT_VERSION < QT_VERSION_CHECK(6, 0, 0)
 	// fallback session management is useless for tdesktop since it doesn't have
diff --git a/Telegram/SourceFiles/core/local_url_handlers.cpp b/Telegram/SourceFiles/core/local_url_handlers.cpp
index 5b13c00f1c..23a3e958e4 100644
--- a/Telegram/SourceFiles/core/local_url_handlers.cpp
+++ b/Telegram/SourceFiles/core/local_url_handlers.cpp
@@ -75,6 +75,7 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "history/history_item.h"
 #include "iv/iv_instance.h"
 #include "apiwrap.h"
+#include "blorg/discovery/discovery.h"
 
 #include "styles/style_chat_helpers.h"
 
@@ -548,6 +549,9 @@ bool ResolveUsernameOrPhone(
 		Window::SessionController *controller,
 		const Match &match,
 		const QVariant &context) {
+	if (!BLORG::Discovery::AllowResolveUsername()) {
+		return false;
+	}
 	if (!controller) {
 		return false;
 	}
@@ -1935,6 +1939,9 @@ QString TryConvertUrlToLocal(QString url) {
 	}
 	auto telegramMeMatch = regex_match(u"^(https?://)?(www\\.)?(telegram\\.(me|dog)|t\\.me)/(.+)$"_q, url, matchOptions);
 	if (telegramMeMatch) {
+		if (!BLORG::Discovery::AllowTelegramLinks()) {
+			return QString();
+		}
 		const auto query = telegramMeMatch->capturedView(5);
 		if (const auto phoneMatch = regex_match(u"^\\+([0-9]+)(\\?|$)"_q, query, matchOptions)) {
 			const auto params = query.mid(phoneMatch->captured(0).size()).toString();
diff --git a/Telegram/SourceFiles/core/update_checker.cpp b/Telegram/SourceFiles/core/update_checker.cpp
index 976c40695c..6e2751a5f4 100644
--- a/Telegram/SourceFiles/core/update_checker.cpp
+++ b/Telegram/SourceFiles/core/update_checker.cpp
@@ -1879,21 +1879,23 @@ bool checkReadyUpdate() {
 
 void UpdateApplication() {
 	if (UpdaterDisabled()) {
-		const auto url = [&] {
-#ifdef OS_WIN_STORE
-			return "https://www.microsoft.com/en-us/store/p/telegram-desktop/9nztwsqntd0s";
-#elif defined OS_MAC_STORE // OS_WIN_STORE
-			return "https://itunes.apple.com/ae/app/telegram-desktop/id946399090";
-#else // OS_WIN_STORE || OS_MAC_STORE
-			if (KSandbox::isFlatpak()) {
-				return "https://flathub.org/apps/details/org.telegram.desktop";
-			} else if (KSandbox::isSnap()) {
-				return "https://snapcraft.io/telegram-desktop";
-			}
-			return "https://desktop.telegram.org";
-#endif // OS_WIN_STORE || OS_MAC_STORE
-		}();
-		UrlClickHandler::Open(url);
+		UrlClickHandler::Open("https://github.com/dmytro-varich/Telegram-BLORG/releases");
+// 	} else 
+// 		const auto url = [&] {
+// #ifdef OS_WIN_STORE
+// 			return "https://www.microsoft.com/en-us/store/p/telegram-desktop/9nztwsqntd0s";
+// #elif defined OS_MAC_STORE // OS_WIN_STORE
+// 			return "https://itunes.apple.com/ae/app/telegram-desktop/id946399090";
+// #else // OS_WIN_STORE || OS_MAC_STORE
+// 			if (KSandbox::isFlatpak()) {
+// 				return "https://flathub.org/apps/details/org.telegram.desktop";
+// 			} else if (KSandbox::isSnap()) {
+// 				return "https://snapcraft.io/telegram-desktop";
+// 			}
+// 			return "https://desktop.telegram.org";
+// #endif // OS_WIN_STORE || OS_MAC_STORE
+// 		}();
+		// UrlClickHandler::Open(url);
 	} else {
 		cSetAutoUpdate(true);
 		const auto window = Core::IsAppLaunched()
diff --git a/Telegram/SourceFiles/core/version.h b/Telegram/SourceFiles/core/version.h
index 4aa6a2a239..f9424d7250 100644
--- a/Telegram/SourceFiles/core/version.h
+++ b/Telegram/SourceFiles/core/version.h
@@ -20,8 +20,8 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 // used in Updater.cpp and Setup.iss for Windows
 constexpr auto AppId = "{53F49750-6209-4FBF-9CA8-7A333C87D1ED}"_cs;
 constexpr auto AppNameOld = "Telegram Win (Unofficial)"_cs;
-constexpr auto AppName = "Telegram Desktop"_cs;
-constexpr auto AppFile = "Telegram"_cs;
+constexpr auto AppName = "Telegram Blorg"_cs;
+constexpr auto AppFile = "TelegramBlorg"_cs;
 constexpr auto AppVersion = 7000009;
 constexpr auto AppVersionStr = "7.0.9";
 constexpr auto AppBetaVersion = false;
diff --git a/Telegram/SourceFiles/dialogs/dialogs_widget.cpp b/Telegram/SourceFiles/dialogs/dialogs_widget.cpp
index 08a51854bd..8d9d8ec947 100644
--- a/Telegram/SourceFiles/dialogs/dialogs_widget.cpp
+++ b/Telegram/SourceFiles/dialogs/dialogs_widget.cpp
@@ -101,6 +101,7 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "styles/style_info.h"
 #include "styles/style_window.h"
 #include "base/qt/qt_common_adapters.h"
+#include "blorg/discovery/discovery.h"
 
 #include <QtCore/QMimeData>
 #include <QtGui/QTextBlock>
@@ -3452,6 +3453,9 @@ void Widget::requestPublicPosts(bool fromStart) {
 }
 
 void Widget::requestMessages(bool fromStart) {
+	if (!BLORG::Discovery::AllowGlobalSearch()) {
+        return; 
+    }
 	if (!_searchProcess.lastId || !_searchProcess.lastPeer) {
 		fromStart = true;
 	}
diff --git a/Telegram/SourceFiles/history/history_item_components.cpp b/Telegram/SourceFiles/history/history_item_components.cpp
index 60f51bec90..460c92f705 100644
--- a/Telegram/SourceFiles/history/history_item_components.cpp
+++ b/Telegram/SourceFiles/history/history_item_components.cpp
@@ -62,6 +62,8 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "styles/style_dialogs.h" // dialogsMiniReplyStory.
 #include "styles/style_widgets.h"
 
+#include "blorg/discovery/discovery.h"
+
 #include <QtGui/QGuiApplication>
 
 namespace {
@@ -383,13 +385,17 @@ void HistoryMessageForwarded::create(
 	}
 	text.setMarkedText(st::fwdTextStyle, phrase, kMarkupTextOptions, context);
 
-	text.setLink(1, fromChannel
-		? JumpToMessageClickHandler(originalSender, originalId)
-		: originalSender
-		? originalSender->openLink()
-		: HiddenSenderInfo::ForwardClickHandler());
-	if (via) {
-		text.setLink(2, via->link);
+	if (BLORG::Discovery::AllowForwardedMessages()) {
+		text.setLink(1, fromChannel
+			? JumpToMessageClickHandler(originalSender, originalId)
+			: originalSender
+			? originalSender->openLink()
+			: HiddenSenderInfo::ForwardClickHandler());
+		if (via) {
+			text.setLink(2, via->link);
+		}
+	} else {
+		text.setLink(1, HiddenSenderInfo::ForwardClickHandler());
 	}
 }
 
diff --git a/Telegram/SourceFiles/platform/linux/specific_linux.cpp b/Telegram/SourceFiles/platform/linux/specific_linux.cpp
index a5249ff47a..03466ba404 100644
--- a/Telegram/SourceFiles/platform/linux/specific_linux.cpp
+++ b/Telegram/SourceFiles/platform/linux/specific_linux.cpp
@@ -232,7 +232,7 @@ bool GenerateDesktopFile(
 	DEBUG_LOG(("App Info: placing .desktop file to %1").arg(targetPath));
 	if (!QDir(targetPath).exists()) QDir().mkpath(targetPath);
 
-	const auto sourceFile = u":/misc/org.telegram.desktop.desktop"_q;
+	const auto sourceFile = u":/misc/blorg.telegram.desktop.desktop"_q;
 	const auto targetFile = targetPath
 		+ QGuiApplication::desktopFileName()
 		+ u".desktop"_q;
@@ -371,7 +371,7 @@ bool GenerateDesktopFile(
 		hashMd5Hex(d.constData(), d.size(), md5Hash);
 
 		if (!Core::Launcher::Instance().customWorkingDir()) {
-			QFile::remove(u"%1org.telegram.desktop._%2.desktop"_q.arg(
+			QFile::remove(u"%1blorg.telegram.desktop._%2.desktop"_q.arg(
 				targetPath,
 				md5Hash));
 
@@ -380,7 +380,7 @@ bool GenerateDesktopFile(
 			hashMd5Hex(exePath.constData(), exePath.size(), md5Hash);
 		}
 
-		QFile::remove(u"%1org.telegram.desktop.%2.desktop"_q.arg(
+		QFile::remove(u"%1blorg.telegram.desktop.%2.desktop"_q.arg(
 			targetPath,
 			md5Hash));
 	}
@@ -439,7 +439,7 @@ bool GenerateServiceFile(bool silent = false) {
 		const auto d = QFile::encodeName(QDir(cWorkingDir()).absolutePath());
 		hashMd5Hex(d.constData(), d.size(), md5Hash);
 
-		QFile::remove(u"%1org.telegram.desktop._%2.service"_q.arg(
+		QFile::remove(u"%1blorg.telegram.desktop._%2.service"_q.arg(
 			targetPath,
 			md5Hash));
 	}
@@ -469,7 +469,7 @@ void InstallLauncher() {
 
 	// don't update desktop file for alpha version or if updater is disabled
 	if (cAlphaVersion()
-			|| Core::UpdaterDisabled()
+			// || Core::UpdaterDisabled()
 			|| KSandbox::isInside()
 			|| DisabledByEnv) {
 		return;
@@ -494,6 +494,14 @@ void InstallLauncher() {
 		DEBUG_LOG(("App Info: Icon copied to '%1'").arg(icon));
 	}
 
+	const auto appIcons512 = icons + u"/hicolor/512x512/apps/"_q;
+	if (!QDir(appIcons512).exists()) QDir().mkpath(appIcons512);
+	QFile::copy(u":/gui/art/logo_256.png"_q, appIcons512 + ApplicationIconName() + u".png"_q);
+
+	QProcess::execute("gtk-update-icon-cache", {
+		"-f", "-t", icons + "hicolor"
+	});
+
 	const auto symbolicIcons = icons + u"/hicolor/symbolic/apps/"_q;
 	if (!QDir().exists(symbolicIcons)) QDir().mkpath(symbolicIcons);
 
@@ -669,16 +677,16 @@ bool SkipTaskbarSupported() {
 }
 
 QString ExecutablePathForShortcuts() {
-	if (Core::UpdaterDisabled()) {
-		const auto &arguments = Core::Launcher::Instance().arguments();
-		if (!arguments.isEmpty()) {
-			const auto result = QFileInfo(arguments.first()).fileName();
-			if (!result.isEmpty()) {
-				return result;
-			}
-		}
-		return cExeName();
-	}
+	// if (Core::UpdaterDisabled()) {
+	// 	const auto &arguments = Core::Launcher::Instance().arguments();
+	// 	if (!arguments.isEmpty()) {
+	// 		const auto result = QFileInfo(arguments.first()).fileName();
+	// 		if (!result.isEmpty()) {
+	// 			return result;
+	// 		}
+	// 	}
+	// 	return cExeName();
+	// }
 	return cExeDir() + cExeName();
 }
 
@@ -737,11 +745,11 @@ void start() {
 		}
 
 		if (!Core::UpdaterDisabled()) {
-			return u"org.telegram.desktop._%1"_q.arg(
+			return u"blorg.telegram.desktop._%1"_q.arg(
 				Core::Launcher::Instance().instanceHash().constData());
 		}
 
-		return u"org.telegram.desktop"_q;
+		return u"blorg.telegram.desktop"_q;
 	}());
 
 	LOG(("App ID: %1").arg(QGuiApplication::desktopFileName()));
@@ -839,11 +847,7 @@ QImage DefaultApplicationIcon() {
 }
 
 QString ApplicationIconName() {
-	static const auto Result = KSandbox::isSnap()
-		? u"snap.%1."_q.arg(qEnvironmentVariable("SNAP_INSTANCE_NAME"))
-		: QGuiApplication::desktopFileName().remove(
-		u"._"_q + Core::Launcher::Instance().instanceHash());
-	return Result;
+	return u"blorg.telegram.desktop"_q;
 }
 
 void LaunchMaps(const Data::LocationPoint &point, Fn<void()> fail) {
diff --git a/Telegram/SourceFiles/platform/win/windows_app_user_model_id.cpp b/Telegram/SourceFiles/platform/win/windows_app_user_model_id.cpp
index 3d051f4ce0..8001a026bf 100644
--- a/Telegram/SourceFiles/platform/win/windows_app_user_model_id.cpp
+++ b/Telegram/SourceFiles/platform/win/windows_app_user_model_id.cpp
@@ -28,7 +28,7 @@ const PROPERTYKEY pkey_AppUserModel_ToastActivator = { { 0x9F4C2855, 0x9F79, 0x4
 #ifdef OS_WIN_STORE
 const WCHAR AppUserModelIdBase[] = L"Telegram.TelegramDesktop.Store";
 #else // OS_WIN_STORE
-const WCHAR AppUserModelIdBase[] = L"Telegram.TelegramDesktop";
+const WCHAR AppUserModelIdBase[] = L"Telegram.TelegramBlorg";
 #endif // OS_WIN_STORE
 
 [[nodiscard]] QString PinnedIconsPath() {
diff --git a/Telegram/SourceFiles/settings/sections/settings_privacy_security.cpp b/Telegram/SourceFiles/settings/sections/settings_privacy_security.cpp
index 2aab5fc974..40a47d0afc 100644
--- a/Telegram/SourceFiles/settings/sections/settings_privacy_security.cpp
+++ b/Telegram/SourceFiles/settings/sections/settings_privacy_security.cpp
@@ -71,6 +71,7 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "styles/style_layers.h"
 #include "styles/style_menu_icons.h"
 #include "styles/style_settings.h"
+#include "blorg/discovery/discovery.h"
 
 #include <QtGui/QGuiApplication>
 #include <QtSvg/QSvgRenderer>
@@ -744,23 +745,26 @@ void BuildSecuritySection(
 		return count ? QString::number(count) : none;
 	});
 
-	builder.addButton({
-		.id = u"security/blocked"_q,
-		.title = tr::lng_settings_blocked_users(),
-		.icon = { &st::menuIconBlock },
-		.label = std::move(blockedCount),
-		.onClick = [=] {
-			showOther(BlockedPeersId());
-		},
-		.keywords = { u"blocked"_q, u"ban"_q },
-	});
+	if (BLORG::Discovery::ShowBlockedUsersSection()) {
+		builder.addButton({
+			.id = u"security/blocked"_q,
+			.title = tr::lng_settings_blocked_users(),
+			.icon = { &st::menuIconBlock },
+			.label = std::move(blockedCount),
+			.onClick = [=] {
+				showOther(BlockedPeersId());
+			},
+			.keywords = { u"blocked"_q, u"ban"_q },
+		});
 
-	builder.add([session, updateTrigger = rpl::duplicate(updateTrigger)](const WidgetContext &ctx) mutable {
-		std::move(updateTrigger) | rpl::on_next([=] {
-			session->api().blockedPeers().reload();
-		}, ctx.container->lifetime());
-		return SectionBuilder::WidgetToAdd{};
-	});
+		builder.add([session, updateTrigger = rpl::duplicate(updateTrigger)](const WidgetContext &ctx) mutable {
+			std::move(updateTrigger) | rpl::on_next([=] {
+				session->api().blockedPeers().reload();
+			}, ctx.container->lifetime());
+			return SectionBuilder::WidgetToAdd{};
+		});
+
+	}
 
 	auto websitesCount = session->api().websites().totalValue();
 	auto websitesShown = rpl::duplicate(websitesCount) | rpl::map(
diff --git a/Telegram/SourceFiles/window/window_session_controller.cpp b/Telegram/SourceFiles/window/window_session_controller.cpp
index 6251bbcc19..a012e7e598 100644
--- a/Telegram/SourceFiles/window/window_session_controller.cpp
+++ b/Telegram/SourceFiles/window/window_session_controller.cpp
@@ -128,6 +128,7 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 #include "styles/style_window.h"
 #include "styles/style_dialogs.h"
 #include "styles/style_layers.h" // st::boxLabel
+#include "blorg/discovery/discovery.h"
 
 namespace Window {
 namespace {
@@ -516,6 +517,10 @@ void SessionNavigation::resolveUsername(
 		const QString &username,
 		Fn<void(not_null<PeerData*>)> done,
 		const QString &referral) {
+	if (!BLORG::Discovery::AllowMentions()) {
+        return;
+    }		
+
 	if (referral.isEmpty()) {
 		if (const auto peer = _session->data().peerByUsername(username)) {
 			done(peer);
diff --git a/Telegram/build/setup.iss b/Telegram/build/setup.iss
index cf3e8d5ff6..c38a3c1596 100644
--- a/Telegram/build/setup.iss
+++ b/Telegram/build/setup.iss
@@ -1,8 +1,8 @@
-#define MyAppShortName "Telegram"
-#define MyAppName "Telegram Desktop"
+#define MyAppShortName "Telegram Blorg"
+#define MyAppName "Telegram Blorg"
 #define MyAppPublisher "Telegram FZ-LLC"
 #define MyAppURL "https://desktop.telegram.org"
-#define MyAppExeName "Telegram.exe"
+#define MyAppExeName "TelegramBlorg.exe"
 #define MyAppId "53F49750-6209-4FBF-9CA8-7A333C87D1ED"
 #define CurrentYear GetDateTimeString('yyyy','','')
 
diff --git a/lib/xdg/org.telegram.desktop.desktop b/lib/xdg/blorg.telegram.desktop.desktop
similarity index 70%
rename from lib/xdg/org.telegram.desktop.desktop
rename to lib/xdg/blorg.telegram.desktop.desktop
index 73ca5cc2d5..305a1f0606 100644
--- a/lib/xdg/org.telegram.desktop.desktop
+++ b/lib/xdg/blorg.telegram.desktop.desktop
@@ -1,11 +1,10 @@
 [Desktop Entry]
-Name=Telegram
-Comment=New era of messaging
-TryExec=Telegram
-Exec=Telegram -- %U
-Icon=org.telegram.desktop
+Name=Telegram Blorg
+Comment=Focus on Work
+Exec=TelegramBlorg -- %U
+Icon=blorg.telegram.desktop
 Terminal=false
-StartupWMClass=TelegramDesktop
+StartupWMClass=TelegramBlorg
 Type=Application
 Categories=Chat;Network;InstantMessaging;Qt;
 MimeType=x-scheme-handler/tg;x-scheme-handler/tonsite;
@@ -17,6 +16,6 @@ X-GNOME-UsesNotifications=true
 X-GNOME-SingleWindow=true
 
 [Desktop Action quit]
-Exec=Telegram -quit
+Exec=TelegramBlorg -quit
 Name=Quit Telegram
 Icon=application-exit
diff --git a/lib/xdg/org.telegram.desktop.metainfo.xml b/lib/xdg/blorg.telegram.desktop.metainfo.xml
similarity index 96%
rename from lib/xdg/org.telegram.desktop.metainfo.xml
rename to lib/xdg/blorg.telegram.desktop.metainfo.xml
index 0fec02f170..398c3c0d50 100644
--- a/lib/xdg/org.telegram.desktop.metainfo.xml
+++ b/lib/xdg/blorg.telegram.desktop.metainfo.xml
@@ -1,10 +1,10 @@
 <?xml version="1.0" encoding="utf-8"?>
 <component type="desktop">
-    <id>org.telegram.desktop</id>
+    <id>blorg.telegram.desktop</id>
     <metadata_license>CC0-1.0</metadata_license>
     <project_license>GPL-3.0</project_license>
-    <name>Telegram</name>
-    <summary>New era of messaging</summary>
+    <name>Telegram Blorg</name>
+    <summary>Focus on Work</summary>
     <description>
         <p>Pure instant messaging — simple, fast, secure, and synced across all your devices. One of the world's top 10 most downloaded apps with over 500 million active users.</p>
         <p>FAST: Telegram is the fastest messaging app on the market, connecting people via a unique, distributed network of data centers around the globe.</p>
@@ -106,10 +106,10 @@
         <content_attribute id="money-gambling">none</content_attribute>
         <content_attribute id="money-advertising">moderate</content_attribute>
     </content_rating>
-    <launchable type="desktop-id">org.telegram.desktop.desktop</launchable>
+    <launchable type="desktop-id">blorg.telegram.desktop</launchable>
     <provides>
-        <binary>Telegram</binary>
-        <dbus type="session">org.telegram.desktop</dbus>
+        <binary>TelegramBlorg</binary>
+        <dbus type="session">blorg.telegram.desktop</dbus>
         <mediatype>x-scheme-handler/tg</mediatype>
         <mediatype>x-scheme-handler/tonsite</mediatype>
     </provides>
diff --git a/lib/xdg/blorg.telegram.desktop.service b/lib/xdg/blorg.telegram.desktop.service
new file mode 100644
index 0000000000..86ba5e4fb9
--- /dev/null
+++ b/lib/xdg/blorg.telegram.desktop.service
@@ -0,0 +1,3 @@
+[D-BUS Service]
+Name=blorg.telegram.desktop
+Exec=@CMAKE_INSTALL_FULL_BINDIR@/TelegramBlorg
\ No newline at end of file
diff --git a/lib/xdg/org.telegram.desktop.service b/lib/xdg/org.telegram.desktop.service
deleted file mode 100644
index 8935a4fa83..0000000000
--- a/lib/xdg/org.telegram.desktop.service
+++ /dev/null
@@ -1,3 +0,0 @@
-[D-BUS Service]
-Name=org.telegram.desktop
-Exec=@CMAKE_INSTALL_FULL_BINDIR@/Telegram
